### DISCUSSION TITLE
[Tensor Core 编程教程]对Warp 指令的疑问

### DISCUSSION THREAD

**Discussion by hbsun2113** [mlc-ai/mlc-zh#125] [Q&A]
我没有任何GPU编程基础，所以问题可能过于基础，请见谅。
在[Tensor Core 编程教程](https://github.com/mlc-ai/notebooks/blob/main/tutorial/tutorial_tensor_core-zh.ipynb)中有如下优化：
![image](https://user-images.githubusercontent.com/27994442/185742674-376da93f-d34c-4464-846d-c2645881b877.png)

我的问题：
1. 对于“使 threadIdx.x extent=32”：
    -  threadIdx.x extent是什么？
    -  我在上图code中没有找到与**32**有关的绑定，请问是如何实现的？
3. “threadIdx.y”和“threadIdx.x”的区别是什么，为什么“warp-level operators”对x有限制，而对y无限制？
4. 对于“确保一个 warp 中的所有线程都可以同时调用 Tensor Core”，请问是如何确保的？

**Comment by Hzfengsy** [comment]
1. threadIdx.x extent是32，通过GPU的其他部分schedule生成的（比如coop fetch），因为单个kernel内threads和blocks数量相同。
2. warp-level operators 要求是一个warp中的线程同时工作，并不会直接对`threadIdx.x`有要求（原文提到是“是决此问题的最简单方法之一”，并非唯一方法）
3. 这个是由cuda编程模型保证的，TVM只负责生成cuda代码，具体后续编译执行由NVIDIA相关软硬件保证
